### PR TITLE
test(zk): add tests for multi proof

### PR DESCRIPTION
**Pull request by dongchangYoo**
In this PR, tests for the proof generation logic for multiple circuits have been added.
During the process, one different logic and a bug were identified and fixed compared to rust-halo2.

### CODE REVIEW

File: tachyon/crypto/commitments/kzg/shplonk_unittest.cc [hunk 1]
```diff
@@ -1,17 +1,138 @@
...
+    }
+  }
+
+  std::vector<PolynomialOpening<Poly>> CreateProverOpenings() const {
+    return base::Map(
+        prover_openings,
+        [](const OwnedPolynomialOpening<Poly, Commitment>& owned_opening) {
+          return PolynomialOpening<Poly>(
+              base::ShallowRef<const Poly>(&owned_opening.poly),
+              base::DeepRef<const typename Poly::Point>(&owned_opening.point),
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/shplonk_unittest.cc:49]: Since it's being repeated, how about adding `typename Point = Poly::Point` above and refactoring this code?

File: tachyon/crypto/commitments/kzg/shplonk_unittest.cc [hunk 2]
```diff
@@ -108,4 +229,29 @@ TEST_F(SHPlonkTest, CreateAndVerifyProof) {
...
+  SimpleTranscriptWriter<Commitment> writer((base::Uint8VectorBuffer()));
+  std::vector<PolynomialOpening<Poly>> prover_openings =
+      owned_openings.CreateProverOpenings();
+  ASSERT_TRUE(pcs_.CreateOpeningProof(prover_openings, &writer));
+
+  base::Buffer read_buf(writer.buffer().buffer(), writer.buffer().buffer_len());
+  SimpleTranscriptReader<Commitment> reader(std::move(read_buf));
+  std::vector<PolynomialOpening<Poly, Commitment>> verifier_openings =
+      owned_openings.CreateVerifyingOpenings();
+
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/shplonk_unittest.cc:252]: How about removing this empty line as done in above for `CreateOpeningProof`?

File: tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h [hunk 3]
```diff
@@ -326,7 +326,8 @@ class CircuitPolynomialBuilder {
   }
 
   void UpdateVanishingLookups(size_t circuit_idx) {
-    size_t num_lookups = committed_lookups_vec_->size();
+    CHECK_LT(circuit_idx, committed_lookups_vec_->size());
+    size_t num_lookups = (*committed_lookups_vec_)[circuit_idx].size();
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h:330]: Can you make this change a separate commit or fix the commit message to include all these changes?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/examples/BUILD.bazel [hunk 4]
```diff
@@ -24,16 +24,46 @@ tachyon_cc_library(
...
+        "//tachyon/zk/plonk/keys:proving_key",
+        "//tachyon/zk/plonk/layout/floor_planner:simple_floor_planner",
+    ],
+)
+
+tachyon_cc_test(
+    name = "simple_lookup_v1_test",
+    srcs = ["simple_lookup_v1_circuit_unittest.cc"],
+    deps = [
+        ":circuit_test",
```
Comment (Insun35) [inline tachyon/zk/plonk/examples/BUILD.bazel:66]: Please sort targets alphabetically. They should be located before `simple_v1_circuit_test`.

**Comment by chokobole** [comment]
https://github.com/kroma-network/tachyon/pull/263/commits/31dc1061d5f683d77d2a1a2ea9bfd3663744de01#diff-6d42394d877ed7c0bac236a68ee117863fc98fd287c560661f1592c77514e489R29 What about using `base::Ref<>` instead of `base::ShallowRef`. By default `SHALLOW` is true.

File: tachyon/crypto/commitments/kzg/shplonk_unittest.cc [hunk 5]
```diff
@@ -1,17 +1,138 @@
...
+        prover_openings,
+        [](const OwnedPolynomialOpening<Poly, Commitment>& owned_opening) {
+          return PolynomialOpening<Poly>(
+              base::ShallowRef<const Poly>(&owned_opening.poly),
+              base::DeepRef<const typename Poly::Point>(&owned_opening.point),
+              owned_opening.opening);
+        });
+  }
+
+  std::vector<PolynomialOpening<Poly, Commitment>> CreateVerifyingOpenings()
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/shplonk_unittest.cc:54]: `CreateProvingOpening()` and `CreateVerifyingOpening()` doesn't correspond to each other.

```suggestion
  std::vector<PolynomialOpening<Poly, Commitment>> CreateVerifierOpenings()
```

File: tachyon/crypto/commitments/kzg/shplonk_unittest.cc [hunk 6]
```diff
@@ -108,4 +229,29 @@ TEST_F(SHPlonkTest, CreateAndVerifyProof) {
   EXPECT_TRUE((pcs_.VerifyOpeningProof(verifier_openings, &reader)));
 }
 
-}  // namespace tachyon::crypto
+TEST_F(SHPlonkTest, ComplexCreateAndVerifyProof) {
+  OwnedPolynomialOpenings<Poly, Commitment> owned_openings;
+  std::string error;
+  EXPECT_TRUE(LoadAndParseJson(
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/shplonk_unittest.cc:235]:
```suggestion
  ASSERT_TRUE(LoadAndParseJson(
```

File: tachyon/zk/plonk/examples/simple_circuit_unittest.cc [hunk 7]
```diff
@@ -660,6 +773,34 @@ TEST_F(SimpleCircuitTest, CreateProof) {
   EXPECT_THAT(proof, testing::ContainerEq(expected_proof));
 }
 
+TEST_F(SimpleCircuitTest, CreateMultiProof) {
```
Comment (chokobole) [inline tachyon/zk/plonk/examples/simple_circuit_unittest.cc:776]: I think then `CreateProof` becomes redundant. What do you think?

File: tachyon/zk/plonk/examples/simple_circuit_unittest.cc [hunk 8]
```diff
@@ -868,4 +1009,264 @@ TEST_F(SimpleCircuitTest, Verify) {
   EXPECT_EQ(h_eval, expected_h_eval);
 }
 
+TEST_F(SimpleCircuitTest, VerifyMultiProof) {
```
Comment (chokobole) [inline tachyon/zk/plonk/examples/simple_circuit_unittest.cc:1012]: I think then `Verify` becomes redundant. What do you think?

File: tachyon/zk/plonk/examples/BUILD.bazel [hunk 9]
```diff
@@ -24,16 +24,46 @@ tachyon_cc_library(
...
 # Change |tachyon_cc_test| to |tachyon_cc_unittest| once CI can run within the timeout.
 tachyon_cc_test(
-    name = "examples_unittests",
-    srcs = [
-        "simple_circuit_unittest.cc",
-        "simple_lookup_circuit_unittest.cc",
-        "simple_lookup_v1_circuit_unittest.cc",
-        "simple_v1_circuit_unittest.cc",
+    name = "simple_circuit_test",
+    srcs = ["simple_circuit_unittest.cc"],
```
Comment (chokobole) [inline tachyon/zk/plonk/examples/BUILD.bazel:28]: Please rename `simple_circuit_unittest.cc` to `simple_circuit_test.cc` ! It looks like integration test.

File: tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h (hunk 3 above)
Comment (dongchangYoo) [reply tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h:330]: @TomTaehoonKim `num_lookups` determines the number of iterations for the following for-loop.

Comment (TomTaehoonKim) [reply tachyon/zk/plonk/vanishing/circuit_polynomial_builder.h]: @chokobole I've checked my comment! You may resolve this after you check it as well.

**Comment by chokobole** [comment]
Please remove scpoe from https://github.com/kroma-network/tachyon/pull/263/commits/e3f185ce88479c44e435654a0b682e9569911e45

**Review by chokobole — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM
